src/sources/kubernetes_logs/pod_metadata_annotator.rs

```diff
@@ -0,0 +1,212 @@
+//! Annotates events with pod metadata.
+
+#![deny(missing_docs)]
+
+use super::{path_helpers::parse_log_file_path, FILE_KEY};
+use crate::{
+    event::{LogEvent, Value},
+    kubernetes as k8s, Event,
+};
+use evmap10::ReadHandle;
+use k8s_openapi::{api::core::v1::Pod, apimachinery::pkg::apis::meta::v1::ObjectMeta};
+use serde::{Deserialize, Serialize};
+use string_cache::Atom;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct FieldsSpec {
+    pub pod_name: String,
+    pub pod_namespace: String,
+    pub pod_uid: String,
+    pub pod_labels: String,
+}
+
+impl Default for FieldsSpec {
+    fn default() -> Self {
+        Self {
+            pod_name: "kubernetes.pod_name".to_owned(),
+            pod_namespace: "kubernetes.pod_namespace".to_owned(),
+            pod_uid: "kubernetes.pod_uid".to_owned(),
+            pod_labels: "kubernetes.pod_labels".to_owned(),
+        }
+    }
+}
+
+/// Annotate the event with pod metadata.
+pub struct PodMetadataAnnotator {
+    pods_state_reader: ReadHandle<String, k8s::state::evmap::Value<Pod>>,
+    fields_spec: FieldsSpec,
+}
+
+impl PodMetadataAnnotator {
+    /// Create a new [`PodMetadataAnnotator`].
+    pub fn new(
+        pods_state_reader: ReadHandle<String, k8s::state::evmap::Value<Pod>>,
+        fields_spec: FieldsSpec,
+    ) -> Self {
+        Self {
+            pods_state_reader,
+            fields_spec,
+        }
+    }
+}
+
+impl PodMetadataAnnotator {
+    /// Annotates an event with the information from the [`Pod::metadata`].
+    /// The event has to be obtained from kubernetes log file, and have a
+    /// [`FILE_KEY`] field set with a file that the line came from.
+    pub fn annotate(&self, event: &mut Event) -> Option<()> {
+        let log = event.as_mut_log();
+        let uid = pod_uid_from_log_event(log)?;
+        let guard = self.pods_state_reader.get(&uid)?;
+        let entry = guard.get_one()?;
+        let pod: &Pod = entry.as_ref();
+        let metadata = pod.metadata.as_ref()?;
+        annotate_from_metadata(log, &self.fields_spec, &metadata);
+        Some(())
+    }
+}
+
+fn pod_uid_from_log_event(log: &LogEvent) -> Option<String> {
+    let value = log.get(&Atom::from(FILE_KEY))?;
+    let file = match value {
+        Value::Bytes(bytes) => String::from_utf8_lossy(bytes).into_owned(),
+        _ => return None,
+    };
+    let info = parse_log_file_path(file.as_str())?;
+    Some(info.pod_uid.to_owned())
+}
+
+fn annotate_from_metadata(log: &mut LogEvent, fields_spec: &FieldsSpec, metadata: &ObjectMeta) {
```

Comment:
This looks like it belongs to `impl FieldSpec{..}`

Reply:
It really belongs to `PodMetadataAnnotator`. Technically, it's enough to bind to `FieldsSpec`, but that relation is unnecessary here, moreover, it can technically be bound to any of it's arguments.
Initially, after I saw this comment I thought it's a good idea, but now I'm a bit hesitant about this change.

Reply:
> It really belongs to PodMetadataAnnotator

Yea, shame that borrow guard is in the way. Good decision, leave it like that.